File: 3.LongestSubstring/wyrj/longestSubstring.js
/* 188ms */
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if ('' === s) {
        return 0;
    }
    var ans = 1;
    var length = 1;
    var begin = 0;
    for (var i = 1; i < s.length; i++) {
        var index = s.substr(begin, length).indexOf(s[i]);
        if (-1 < index) {
            index = index + begin;
            begin = index + 1;
            length = i - index;
        } else {
            length++;
            if (length > ans) {
                ans = length;
            }
        }
    }
    return ans;
};

File: 3.LongestSubstring/wyrj/longestSubstring.c
/* 8ms */
int findIndex(char* s, int begin, int end) {
    int i;
    for (i = begin; i < end; i++) {
        if (s[i] == s[end]) {
            return i;
        }
    }
    return -1;
}
int lengthOfLongestSubstring(char* s) {
    int length = strlen(s);
    if (0 == length) {
        return 0;
    }
    int ans = 1, nowLen = 1, begin = 0, i;
    for (i = 1; i < length; i++) {
        int index = findIndex(s, begin, i);
        if (-1 < index) {
            nowLen = i - index;
            begin = index + 1;
        } else {
            nowLen++;
            if (nowLen > ans) {
                ans = nowLen;
            }
        }
    }
    return ans;
}